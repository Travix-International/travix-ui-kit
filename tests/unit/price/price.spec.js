import { shallow } from 'enzyme';
import React from 'react';
import Price, { addThousandsSeparator, ensureDecimalPrecision } from '../../../components/price/price';

describe('Price', () => {
  describe('#addThousandsSeparator', () => {
    it('returns original value when thousandsSeparator is undefined || value.length < 3', () => {
      expect(addThousandsSeparator('5000')).toEqual('5000');
      expect(addThousandsSeparator('500', '.')).toEqual('500');
    });

    it('returns value with given thousandsSeparator', () => {
      expect(addThousandsSeparator('5000', '.')).toEqual('5.000');
      expect(addThousandsSeparator('50000', '.')).toEqual('50.000');
      expect(addThousandsSeparator('50000', ' ')).toEqual('50 000');
      expect(addThousandsSeparator('5000000', ' ')).toEqual('5 000 000');
    });
  });

  describe('#ensureDecimalPrecision', () => {
    it('returns truncated value to decimalsPrecision number when value.length > decimalsPrecision', () => {
      expect(ensureDecimalPrecision('5000')).toEqual('50');
      expect(ensureDecimalPrecision('5000', 3)).toEqual('500');
    });

    it('returns value suffixed by 0s value.length < decimalsPrecision until matches decimalsPrecision length', () => {
      expect(ensureDecimalPrecision('5')).toEqual('50');
      expect(ensureDecimalPrecision('5', 3)).toEqual('500');
      expect(ensureDecimalPrecision('', 3)).toEqual('000');
      expect(ensureDecimalPrecision(undefined, 3)).toEqual('000');
    });
  });

  describe('#render()', () => {
    it('should return base class with mods for strings as class and string as mods', () => {
      const dataAttrs = {
        'gtm-id': 'test',
      };

      const wrapper = shallow(
        <Price
          dataAttrs={dataAttrs}
          decimalsSeparator=","
          mods={['my-special-class']}
          symbol="د.إ.‏"
          symbolPosition="right"
          thousandsSeparator="."
          underlined
          value={50153.30}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should not render underline when underlined flag is missing', () => {
      const dataAttrs = {
        'gtm-id': 'test',
      };

      const wrapper = shallow(
        <Price
          dataAttrs={dataAttrs}
          decimalsSeparator=","
          mods={['my-special-class']}
          symbol="د.إ.‏"
          symbolPosition="right"
          thousandsSeparator="."
          value={50153.30}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should render only the amount of decimals defined in the decimalsPrecision', () => {
      const wrapper = shallow(
        <Price
          decimalsSeparator=","
          symbol="د.إ.‏"
          symbolPosition="right"
          thousandsSeparator="."
          value={50153.30121212121}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should render without thousands separator when overriden to empty string', () => {
      const wrapper = shallow(
        <Price
          decimalsSeparator=","
          symbol="د.إ.‏"
          symbolPosition="right"
          thousandsSeparator=""
          value={50153.30}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should render without decimals when showDecimals = false', () => {
      const wrapper = shallow(
        <Price
          decimalsSeparator=","
          mods={['my-special-class']}
          showDecimals={false}
          symbol="د.إ.‏"
          symbolPosition="right"
          thousandsSeparator="."
          underlined
          value={50153.30}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should render asterisk when showAsterisk is true', () => {
      const wrapper = shallow(
        <Price
          decimalsSeparator=","
          showAsterisk
          thousandsSeparator="."
          underlined
          value={50153.30}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should render with ui-price_size_xl class', () => {
      const wrapper = shallow(
        <Price
          size="xl"
          value={50153.30}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should render with decimals when value is integer', () => {
      const wrapper = shallow(
        <Price value={50153} />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should render with discount when value is integer not equal 0', () => {
      const wrapper = shallow(
        <Price
          discount={50200}
          value={50153}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should render with additional text if it is non empty string', () => {
      const wrapper = shallow(
        <Price
          additionalText={'per day'}
          value={50153}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should render with additional text block if it is a node', () => {
      const additionalText = (
        <div className="custom-class">
          <span> some important info about price! </span>
        </div>
      );

      const wrapper = shallow(
        <Price
          additionalText={additionalText}
          value={50153}
        />
      );

      expect(wrapper).toMatchSnapshot();
    });

    it('should return null if value is null', () => {
      const wrapper = shallow(
        <Price
          additionalText={'per day'}
          value={null}
        />
      );

      expect(wrapper.html()).toEqual(null);
    });

    it('should return null if value is undefined', () => {
      const wrapper = shallow(
        <Price
          additionalText={'per day'}
          value={undefined}
        />
      );

      expect(wrapper.html()).toEqual(null);
    });

    it('should render unstyled block if an appropriate prop is passed', () => {
      const wrapper = shallow(<Price unstyled value={50153.123} />);

      expect(wrapper).toMatchSnapshot();
    });

    it('should render unstyled block with no decimals', () => {
      const wrapper = shallow(<Price showDecimals={false} unstyled value={50153.123} />);

      expect(wrapper).toMatchSnapshot();
    });

    it('should render unstyled block with right symbol position', () => {
      const wrapper = shallow(<Price symbolPosition="right" unstyled value={50153.123} />);

      expect(wrapper).toMatchSnapshot();
    });

    it('should render unstyled block with asterisk', () => {
      const wrapper = shallow(<Price showAsterisk unstyled value={50153.123} />);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
