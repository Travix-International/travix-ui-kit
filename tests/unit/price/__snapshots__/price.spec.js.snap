// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price #render() should not render underline when underlined flag is missing 1`] = `
<div
  className="ui-price ui-price_my-special-class ui-price_size_l"
  data-gtm-id="test"
>
  <div
    className="ui-price__value-delimiter"
  >
    <div
      className="ui-price__currency ui-price__currency_right"
    >
      د.إ.‏
    </div>
    <div
      className="ui-price__integers"
    >
      50.153
    </div>
    <div
      className="ui-price__decimals"
    >
      ,30
    </div>
  </div>
</div>
`;

exports[`Price #render() should render asterisk when showAsterisk is true 1`] = `
<div
  className="ui-price ui-price_size_l"
>
  <div
    className="ui-price__value-delimiter"
  >
    <div
      className="ui-price__currency ui-price__currency_left"
    >
      €
    </div>
    <div
      className="ui-price__integers"
    >
      50.153
    </div>
    <div
      className="ui-price__decimals"
    >
      ,30
    </div>
    <div
      className="ui-price__asterisk"
    >
      *
    </div>
  </div>
  <div
    className="ui-price__underline"
  />
</div>
`;

exports[`Price #render() should render only the amount of decimals defined in the decimalsPrecision 1`] = `
<div
  className="ui-price ui-price_size_l"
>
  <div
    className="ui-price__value-delimiter"
  >
    <div
      className="ui-price__currency ui-price__currency_right"
    >
      د.إ.‏
    </div>
    <div
      className="ui-price__integers"
    >
      50.153
    </div>
    <div
      className="ui-price__decimals"
    >
      ,30
    </div>
  </div>
</div>
`;

exports[`Price #render() should render unstyled block if an appropriate prop is passed 1`] = `
<span>
  € 50,153.12
</span>
`;

exports[`Price #render() should render unstyled block with asterisk 1`] = `
<span>
  € 50,153.12 *
</span>
`;

exports[`Price #render() should render unstyled block with no decimals 1`] = `
<span>
  € 50,153
</span>
`;

exports[`Price #render() should render unstyled block with right symbol position 1`] = `
<span>
  50,153.12 €
</span>
`;

exports[`Price #render() should render with additional text block if it is a node 1`] = `
<div
  className="ui-price ui-price_size_l"
>
  <div
    className="ui-price__value-delimiter"
  >
    <div
      className="ui-price__currency ui-price__currency_left"
    >
      €
    </div>
    <div
      className="ui-price__integers"
    >
      50,153
    </div>
    <div
      className="ui-price__decimals"
    >
      .00
    </div>
  </div>
  <div
    className="ui-price__additional-text-block"
  >
    <div
      className="custom-class"
    >
      <span>
         some important info about price! 
      </span>
    </div>
  </div>
</div>
`;

exports[`Price #render() should render with additional text if it is non empty string 1`] = `
<div
  className="ui-price ui-price_size_l"
>
  <div
    className="ui-price__value-delimiter"
  >
    <div
      className="ui-price__currency ui-price__currency_left"
    >
      €
    </div>
    <div
      className="ui-price__integers"
    >
      50,153
    </div>
    <div
      className="ui-price__decimals"
    >
      .00
    </div>
  </div>
  <div
    className="ui-price__additional-text-block"
  >
    per day
  </div>
</div>
`;

exports[`Price #render() should render with decimals when value is integer 1`] = `
<div
  className="ui-price ui-price_size_l"
>
  <div
    className="ui-price__value-delimiter"
  >
    <div
      className="ui-price__currency ui-price__currency_left"
    >
      €
    </div>
    <div
      className="ui-price__integers"
    >
      50,153
    </div>
    <div
      className="ui-price__decimals"
    >
      .00
    </div>
  </div>
</div>
`;

exports[`Price #render() should render with discount when value is integer not equal 0 1`] = `
<div
  className="ui-price ui-price_size_l"
>
  <div
    className="ui-price__discount"
  >
    50,200.00
  </div>
  <div
    className="ui-price__value-delimiter"
  >
    <div
      className="ui-price__currency ui-price__currency_left"
    >
      €
    </div>
    <div
      className="ui-price__integers"
    >
      50,153
    </div>
    <div
      className="ui-price__decimals"
    >
      .00
    </div>
  </div>
</div>
`;

exports[`Price #render() should render with ui-price_size_xl class 1`] = `
<div
  className="ui-price ui-price_size_xl"
>
  <div
    className="ui-price__value-delimiter"
  >
    <div
      className="ui-price__currency ui-price__currency_left"
    >
      €
    </div>
    <div
      className="ui-price__integers"
    >
      50,153
    </div>
    <div
      className="ui-price__decimals"
    >
      .30
    </div>
  </div>
</div>
`;

exports[`Price #render() should render without decimals when showDecimals = false 1`] = `
<div
  className="ui-price ui-price_my-special-class ui-price_size_l"
>
  <div
    className="ui-price__value-delimiter"
  >
    <div
      className="ui-price__currency ui-price__currency_right"
    >
      د.إ.‏
    </div>
    <div
      className="ui-price__integers"
    >
      50.153
    </div>
  </div>
  <div
    className="ui-price__underline"
  />
</div>
`;

exports[`Price #render() should render without thousands separator when overriden to empty string 1`] = `
<div
  className="ui-price ui-price_size_l"
>
  <div
    className="ui-price__value-delimiter"
  >
    <div
      className="ui-price__currency ui-price__currency_right"
    >
      د.إ.‏
    </div>
    <div
      className="ui-price__integers"
    >
      50153
    </div>
    <div
      className="ui-price__decimals"
    >
      ,30
    </div>
  </div>
</div>
`;

exports[`Price #render() should return base class with mods for strings as class and string as mods 1`] = `
<div
  className="ui-price ui-price_my-special-class ui-price_size_l"
  data-gtm-id="test"
>
  <div
    className="ui-price__value-delimiter"
  >
    <div
      className="ui-price__currency ui-price__currency_right"
    >
      د.إ.‏
    </div>
    <div
      className="ui-price__integers"
    >
      50.153
    </div>
    <div
      className="ui-price__decimals"
    >
      ,30
    </div>
  </div>
  <div
    className="ui-price__underline"
  />
</div>
`;
